### PR TITLE
copr: optimisze json

### PR DESCRIPTION
**Pull request by BusyJay**
Reduce allocation and add more test cases.

### CODE REVIEW

**Comment by ngaut** [comment]
Could you add some benchmarks ?

File: src/coprocessor/dag/expr/json.rs [hunk 1]
```diff
@@ -156,59 +163,29 @@ impl<'a> JsonFuncArgsParser<'a> {
...
-    where
-        It: Iterator<Item = &'b Expression>,
-    {
-        let func = |e: &'b Expression| {
-            let bytes = try_opt!(e.eval_string(self.ctx, self.row)).into_owned();
-            String::from_utf8(bytes).map(Some).map_err(Error::from)
-        };
-        JsonFuncArgsParser::parse(args, func)
+    fn get_string(&self, e: &Expression) -> Result<Option<String>> {
+        let bytes = try_opt!(e.eval_string(self.ctx, self.row)).into_owned();
```
Comment (hicqu) [inline src/coprocessor/dag/expr/json.rs]: Is it a bug? Should we call `eval_string_and_decode` anywhere we need a `String`?

**Comment by hicqu** [comment]
@ngaut I will add benchmarks about JSON in next PR, and do some tuning.

**Comment by hicqu** [comment]
LGTM.

**Comment by BusyJay** [comment]
@AndreMouche @disksing PTAL

File: src/coprocessor/dag/expr/json.rs [hunk 2]
```diff
@@ -124,9 +124,12 @@ impl FnCall {
         row: &'a [Datum],
     ) -> Result<Option<Cow<'a, Json>>> {
         let parser = JsonFuncArgsParser::new(ctx, row);
-        let head = try_opt!(self.children[0].eval_json(ctx, row)).into_owned();
-        let suffixes = try_opt!(parser.get_jsons_not_none(self.children[1..].iter()));
-        Ok(Some(Cow::Owned(head.merge(suffixes))))
+        let mut head = try_opt!(self.children[0].eval_json(ctx, row)).into_owned();
+        for e in &self.children[1..] {
+            let suffix = try_opt!(parser.get_json_not_none(e));
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/json.rs]: Could we  check none before any `merge`?

File: src/coprocessor/codec/mysql/json/json_merge.rs [hunk 3]
```diff
@@ -21,47 +21,40 @@ impl Json {
...
-                        obj.insert(sub_key, v);
-                    }
-                    Json::Object(obj)
-                }
-                // rule 4
-                (obj, Json::Array(mut sub_array)) => {
-                    let mut array = vec![obj];
-                    array.append(&mut sub_array);
-                    Json::Array(array)
+    pub fn merge(self, to_merge: Json) -> Json {
```
Comment (AndreMouche) [inline src/coprocessor/codec/mysql/json/json_merge.rs]: The annotation on L19 seems should be updated.

File: src/coprocessor/dag/expr/json.rs [hunk 4]
```diff
@@ -57,11 +57,11 @@ impl FnCall {
...
-        if !self.children.is_empty() {
-            let parser = JsonFuncArgsParser::new(ctx, row);
-            let keys = try_opt!(parser.get_strings(self.children.iter().step_by(2)));
-            let elems = try_opt!(parser.get_jsons(self.children[1..].iter().step_by(2)));
-            pairs.extend(keys.into_iter().zip(elems.into_iter()));
+        let parser = JsonFuncArgsParser::new(ctx, row);
+        for chunk in self.children.chunks(2) {
+            let key = try_opt!(chunk[0].eval_string_and_decode(ctx, row)).into_owned();
+            let val = try_opt!(parser.get_json(&chunk[1]));
+            pairs.insert(key, val);
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/json.rs]: Could we check None before any `insert`?

File: src/coprocessor/select/xeval/evaluator.rs [hunk 5]
```diff
@@ -567,10 +567,21 @@ impl Evaluator {
         if children.iter().any(|item| *item == Datum::Null) {
             return Ok(Datum::Null);
         }
+        let l = children.len();
+        if l < 2 {
```
Comment (AndreMouche) [inline src/coprocessor/select/xeval/evaluator.rs]: It has been checked on L566.

File: src/coprocessor/dag/expr/json.rs (hunk 4 above)
Comment (BusyJay) [reply src/coprocessor/dag/expr/json.rs]: How?

**Pull request merged**